src/metabase/db/data_source.clj

```diff
@@ -10,6 +12,61 @@
   (:import java.sql.DriverManager
            java.util.Properties))
 
+(defn- h2? [url]
+  (.startsWith url "jdbc:h2:"))
+
+(defn- h2-path [url]
+  (let [[_ path] (re-matches #"jdbc:h2:file:(.*)$" url)]
+    path))
+
+(defn- h2-migration-paths [url]
+  (let [base (h2-path url)]
+    {:v1db   (str base ".mv.db")
+     :v2db   (str base ".v2.mv.db")
+     :script (str base ".v2migration.sql")}))
+
+(defn- h2-v1-db [url]
+  (let [src (.newInstance (Class/forName "org.h2_v1_4_197.jdbcx.JdbcDataSource"))]
+    (.setURL src url)
+    (.getConnection src)))
+
+(defn- h2-v2-db [url]
+  (let [src (.newInstance (Class/forName "org.h2.jdbcx.JdbcDataSource"))]
+    (.setURL src (str url ".v2"))
+    (.getConnection src)))
+
+(defn- h2-migrate! [url]
+  (let [{:keys [v2db script]} (h2-migration-paths url)]
+    (try
+      (log/warn "H2 migration: beginning migration")
+      (jdbc/query {:connection (h2-v1-db url)} ["SCRIPT TO ?" script])
+      (log/warn "H2 migration: v1 export complete, starting v2 import")
+      (let [conn-v2          (h2-v2-db url)]
+        (jdbc/execute! {:connection conn-v2} ["RUNSCRIPT FROM ? FROM_1X" script])
+        (log/warn "H2 migration: complete")
+        conn-v2)
+      (catch Exception e
+        (log/error "H2 migration failed: " e)
```

Comment:
Should a migration failure kill Metabase?

I'd argue yes, since an error during import can lead to a scary but recoverable UX: getting the setup wizard and thinking you've lost all your Metabase setup!

Reply:
Yeah I think that makes sense